eth/downloader/skeleton.go

```diff
@@ -1120,26 +1120,41 @@ func (s *skeleton) cleanStales(filled *types.Header) error {
 	number := filled.Number.Uint64()
 	log.Trace("Cleaning stale beacon headers", "filled", number, "hash", filled.Hash())
 
-	// If the filled header is below the linked subchain, something's
-	// corrupted internally. Report and error and refuse to do anything.
-	if number < s.progress.Subchains[0].Tail {
+	// If the filled header is below and discontinuous with the linked subchain,
+	// something's corrupted internally. Report and error and refuse to do anything.
+	if number+1 < s.progress.Subchains[0].Tail {
 		return fmt.Errorf("filled header below beacon header tail: %d < %d", number, s.progress.Subchains[0].Tail)
 	}
+	// Determine the new tail based on the given filled header. If it's still
+	// in the range of skeleton chain, use the next header adjacent to the
+	// filled one (it's expected to link with the filled one, otherwise
+	// something corrupted); otherwise use the filled one if the whole skeleton
+	// chain is consumed.
+	newTail := filled
+	if number < s.progress.Subchains[0].Head {
+		newTail = rawdb.ReadSkeletonHeader(s.db, number+1)
+		if newTail.ParentHash != filled.Hash() {
+			return fmt.Errorf("filled header is discontinuous with subchain: %d %s", number, filled.Hash())
+		}
+	}
 	// Subchain seems trimmable, push the tail forward up to the last
 	// filled header and delete everything before it - if available. In
 	// case we filled past the head, recreate the subchain with a new
 	// head to keep it consistent with the data on disk.
 	var (
 		start = s.progress.Subchains[0].Tail // start deleting from the first known header
-		end   = number                       // delete until the requested threshold
+		end   = newTail.Number.Uint64()      // delete skeleton headers before the new tail
 		batch = s.db.NewBatch()
 	)
-	s.progress.Subchains[0].Tail = number
```

Comment:
Ah hmm, indeed this is wonky. If the tail is the first "missing" header, then it should be filled+1